usb/ehci/ehci.h

```diff
@@ -96,154 +139,59 @@ enum { usb_otg1_ctrl = 0x200, usb_otg2_ctrl, usb_otg1_phy_ctrl = usb_otg2_ctrl +
 enum { ehci_item_itd = 0, ehci_item_qh, ehci_item_sitd, ehci_item_fstn };
 
 
-typedef struct {
-	uint32_t terminate : 1;
-	uint32_t type : 2;
-	uint32_t zero : 2;
-	uint32_t pointer : 27;
-} __attribute__((__packed__)) link_pointer_t;
-
-
-struct itd {
-	link_pointer_t next;
-
-	union {
-		struct {
-			uint32_t offset : 12;
-			uint32_t pageSelect : 3;
-			uint32_t ioc : 1;
-			uint32_t length : 12;
-			uint32_t status : 4;
-		};
-		uint32_t raw;
-	} transactions[8];
-
-	union {
-		struct {
-			uint32_t reserved : 12;
-			uint32_t pointer : 20;
-		} buffers[7];
-
-		struct {
-			uint32_t devAddress : 7;
-			uint32_t reserved : 1;
-			uint32_t ep : 4;
-			uint32_t page0 : 20;
-			uint32_t maxPacketSize : 11;
-			uint32_t direction : 1;
-			uint32_t page1 : 20;
-			uint32_t mult : 2;
-		};
-	};
-} __attribute__((__packed__));
-
-
-struct sitd {
-	link_pointer_t next;
-
-	uint32_t devAddr : 7;
-	uint32_t reserved0 : 1;
-	uint32_t ep : 4;
-	uint32_t reserved1 : 4;
-	uint32_t hubAddr : 7;
-	uint32_t reserved2 : 1;
-	uint32_t portNumber : 7;
-	uint32_t direction : 1;
-
-	uint32_t splitStartMask : 8;
-	uint32_t splitCompletionMask : 8;
-	uint32_t reserved3 : 16;
-
-	uint32_t status : 8;
-	uint32_t splitProgressMask : 8;
-	uint32_t bytesToTransfer : 10;
-	uint32_t reserved4 : 4;
-	uint32_t pageSelect : 1;
-	uint32_t ioc : 1;
-
-	uint32_t currentOffset : 12;
-	uint32_t page0 : 20;
-
-	uint32_t transactionCount : 3;
-	uint32_t transactionPosition : 2;
-	uint32_t reserved5 : 7;
-	uint32_t page1 : 20;
-
-	link_pointer_t backLink;
-} __attribute__((__packed__));
-
-
 struct qtd {
-	link_pointer_t next;
-	link_pointer_t altNext;
-
-	uint32_t pingState : 1;
-	uint32_t splitState : 1;
-	uint32_t missedUframe : 1;
-	uint32_t transactionError : 1;
-	uint32_t babble : 1;
-	uint32_t bufferError : 1;
-	uint32_t halted : 1;
-	uint32_t active : 1;
-
-	uint32_t pid : 2;
-	uint32_t errorCounter : 2;
-	uint32_t currentPage : 3;
-	uint32_t ioc : 1;
-	uint32_t bytesToTransfer : 15;
-	uint32_t dt : 1;
-
-	union {
-		struct {
-			uint32_t reserved : 12;
-			uint32_t page : 20;
-		} buffers[5];
-
-		struct {
-			uint32_t offset : 12;
-			uint32_t page0 : 20;
-		};
-	};
-} __attribute__((__packed__));
+	uint32_t next;
+	uint32_t altnext;
+	uint32_t token;
+	uint32_t buf[5];
+};
 
 
 struct qh {
-	link_pointer_t horizontal;
+	uint32_t horizontal;
+	uint32_t info[2];
+	uint32_t current;
+
+	/* Overlay area */
+	uint32_t nextQtd;
+	uint32_t altnextQtd;
+	uint32_t token;
+	uint32_t buf[5];
+};
+
 
-	uint32_t devAddr : 7;
-	uint32_t inactivate : 1;
-	uint32_t ep : 4;
-	uint32_t epSpeed : 2;
-	uint32_t dt : 1;
-	uint32_t headOfReclamation : 1;
-	uint32_t maxPacketLen : 11;
-	uint32_t ctrlEp : 1;
-	uint32_t nakCountReload : 4;
+typedef struct _ehci_qtd {
+	struct _ehci_qtd *prev, *next;
+	volatile struct qtd *hw;
```

Comment:
hw is volatile or pointer to hw is volatile? Or both? Just making sure. Maybe:
```suggestion
	volatile struct qtd * volatile hw;
```

Reply:
The structure to which `hw` points is volatile, as it is accessed by the host controller.